final/master.py

```diff
@@ -0,0 +1,408 @@
+    simplify_strength = 255 / 3
+
+    for x in range(0, WIDTH - 1):
+        for y in range(0, HEIGHT - 1):
+            red = window.get_at((x, y)).r
+            green = window.get_at((x, y)).g
+            blue = window.get_at((x, y)).b
+
+            # Essentially this divides by a number, truncates the decimal portion then multiplies it back to scale
+            # for instance were the strength = 100, a value of 234 would go to 2.34 to 2 then return 200
+            new_red = clamp(simplify_strength * int(red / simplify_strength) + (simplify_strength / 2))
+            new_green = clamp(simplify_strength * int(green / simplify_strength) + (simplify_strength / 2))
+            new_blue = clamp(simplify_strength * int(blue / simplify_strength) + (simplify_strength / 2))
+
+            px_array[x, y] = (new_red, new_green, new_blue)
+
+
+def pixelise_is_a_word_stop_underlining_it_in_grey():
+
+    tile_size = 20
+    tile_spacing = 5
+    global px_array
+    pixel_array = pygame.PixelArray(window)
+
+    for currentStartX in range(0, WIDTH, tile_size):
+        for currentStartY in range(0, HEIGHT, tile_size):
+
+            red_total = 0
+            green_total = 0
+            blue_total = 0
+            colour_count = 0
+
+            for x in range(currentStartX, currentStartX + tile_size - 1):
+                for y in range(currentStartY, currentStartY + tile_size - 1):
+                    current_red = window.get_at((x, y)).r
+                    current_green = window.get_at((x, y)).g
+                    current_blue = window.get_at((x, y)).b
+
+                    red_total += current_red
+                    green_total += current_green
+                    blue_total += current_blue
+                    colour_count += 1
+
+            # Have to swap red and blue here sometimes... No idea why. Maybe png and jpg render bgr not rgb
+            new_red = red_total / colour_count
+            new_green = green_total / colour_count
+            new_blue = blue_total / colour_count
+
+            pixel_array[currentStartX: currentStartX + tile_size - tile_spacing,
+                currentStartY: currentStartY + tile_size - tile_spacing] = (new_red, new_green, new_blue)
+
+    pict = pixel_array.make_surface()
+    del pixel_array
+    del px_array
+
+    window.blit(pict, (0, 0))
+    pygame.display.update()
+
+
+def print_controls():
+    print ('Controls:')
+    print ('Q - Invert colours')
+    print ('W - Blur')
+    print ('E - Night vision')
+    print ('R - Waterfall')
+    print ('T - Horizontal lines, length is similarity strength')
+    print ('Y - Circles')
+    print ('U - Outline')
+    print ('I - Horizontal lines, fading, chosen by strength')
+    print ('O - Simplify colours')
+    print ('P - Pixelise, with tile effect')
+    print ('L - P then O combo. Works well.')
+    print ('Space - Reset')
+
+print_controls()
+
+
+while True:
+    keys = pygame.key.get_pressed()
+    px_array = pygame.PixelArray(window)
```

Comment:
This `PixelArray` should be created and destroyed in each iteration of this loop, and passed into the various functions rather than relying on global declarations.